# Import the repository from GitHub

First of all we start by importing the repository that we stored in the github project.

In [1]:
from kaggle_secrets import UserSecretsClient
user_secrets = UserSecretsClient()
secret_value_1 = user_secrets.get_secret("NEW_GITHUB_TOKEN")

In [2]:
token = UserSecretsClient().get_secret("NEW_GITHUB_TOKEN")
! git clone https://{token}@github.com/madratak/DIQ_Project2024.git

Cloning into 'DIQ_Project2024'...
remote: Enumerating objects: 125, done.
remote: Counting objects: 100% (125/125), done.
remote: Compressing objects: 100% (107/107), done.
remote: Total 125 (delta 53), reused 33 (delta 7), pack-reused 0 (from 0)
Receiving objects: 100% (125/125), 4.40 MiB | 17.26 MiB/s, done.
Resolving deltas: 100% (53/53), done.


# Set up the dataset

At this point we can proceed by importing the correct libraries and then importing the data itself inside our notebook.

In [3]:
# to compute the edit distance
!pip install rapidfuzz 

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 35.9 MB/s eta 0:00:00


In [4]:
import pandas as pd
import numpy as np
from datetime import datetime
import os
import re
from rapidfuzz import fuzz
from tqdm import tqdm
from sklearn.feature_extraction.text import CountVectorizer
from scipy.spatial.distance import jaccard

# Set max column width to None to display full content
pd.set_option('display.max_colwidth', None)

In [5]:
SERVICES = pd.read_csv('/kaggle/working/DIQ_Project2024/data/raw/Comune-di-Milano-Servizi-alla-persona-parrucchieri-estetisti.csv',sep=';',encoding='unicode_escape')
SERVICES.head()

Tipo esercizio pa                           Ubicazione Tipo via  \
0               NaN     LGO DEI GELSOMINI N. 10 (z.d. 6)      LGO   
1               NaN              PZA FIDIA N. 3 (z.d. 9)      PZA   
2               NaN             VIA ADIGE N. 10 (z.d. 5)      VIA   
3               NaN  VIA BARACCHINI FLAVIO N. 9 (z.d. 1)      VIA   
4               NaN           VIA BERGAMO N. 12 (z.d. 4)      VIA   

                 Via Civico  Codice via ZD  \
0      DEI GELSOMINI     10      5394.0  6   
1              FIDIA      3      1144.0  9   
2              ADIGE     10      4216.0  5   
3  BARACCHINI FLAVIO      9       356.0  1   
4            BERGAMO     12      3189.0  4   

                                Prevalente  Superficie altri usi  \
0                                      NaN                   NaN   
1  CENTRO MASSAGGI RILASSANTI NON ESTETICI                   2.0   
2                         CENTRO BENESSERE                   2.0   
3                    TRUCCO SEMIPERMANENTE                   NaN   
4                                      NaN                   NaN   

   Superficie lavorativa  
0                   55.0  
1                   28.0  
2                   27.0  
3                    NaN  
4                   50.0

# Import external dataset
During the data cleaning stage, external are used to fix irregular values, among these we have

In [6]:
viario = pd.read_csv('/kaggle/working/DIQ_Project2024/data/external/VIARIO_20241104.csv',sep=';',encoding='unicode_escape')
centroidi = pd.read_csv('/kaggle/working/DIQ_Project2024/data/external/CENTROIDI_20241104.csv',sep=';',encoding='unicode_escape')
tipovia = pd.read_csv('/kaggle/working/DIQ_Project2024/data/external/TIPOVIA.csv',sep=';',encoding='unicode_escape')
viario_centroidi = viario.merge(centroidi, on='CODICE_VIA', how='left')

civici = pd.read_csv('/kaggle/working/DIQ_Project2024/data/external/CiviciCdMI.csv',sep=';',encoding='unicode_escape')

# Wrangling 

This script focuses on renaming the columns of the `SERVICES` dataset to shorter, more convenient names.  

The original DataFrame is modified in place, and a copy of the updated dataset, named `new_SERVICES`, is created for further analysis or processing.  

We pass from:

| Tipo esercizio pa     | Ubicazione     | Tipo via | Via      | Civico | Codice via | ZD | Superficie altri usi | Superficie lavorativa |
|-------|------|-------|-----|-----|---------|----|---------|---------|

to:


| t_es  | ubic | t_via | via | civ | cod_via | zd | sup_alt | sup_lav | 
|-------|------|-------|-----|-----|---------|----|---------|---------|


In [7]:
new_SERVICES = SERVICES.copy()

new_SERVICES = new_SERVICES.rename(columns={
    'Tipo esercizio pa': 't_es',
    'Ubicazione': 'ubic',
    'Tipo via': 't_via',
    'Via': 'via',
    'Civico': 'civ',
    'Codice via': 'cod_via',
    'ZD': 'zd',
    'Prevalente': 'main',
    'Superficie altri usi': 'sup_alt',
    'Superficie lavorativa': 'sup_lav'
})

new_SERVICES.head()

t_es                                 ubic t_via                via civ  \
0  NaN     LGO DEI GELSOMINI N. 10 (z.d. 6)   LGO      DEI GELSOMINI  10   
1  NaN              PZA FIDIA N. 3 (z.d. 9)   PZA              FIDIA   3   
2  NaN             VIA ADIGE N. 10 (z.d. 5)   VIA              ADIGE  10   
3  NaN  VIA BARACCHINI FLAVIO N. 9 (z.d. 1)   VIA  BARACCHINI FLAVIO   9   
4  NaN           VIA BERGAMO N. 12 (z.d. 4)   VIA            BERGAMO  12   

   cod_via zd                                     main  sup_alt  sup_lav  
0   5394.0  6                                      NaN      NaN     55.0  
1   1144.0  9  CENTRO MASSAGGI RILASSANTI NON ESTETICI      2.0     28.0  
2   4216.0  5                         CENTRO BENESSERE      2.0     27.0  
3    356.0  1                    TRUCCO SEMIPERMANENTE      NaN      NaN  
4   3189.0  4                                      NaN      NaN     50.0

# Transformation
## ZD: Removing data representing different information

This script processes the `zd` column in the `SERVICES` dataset. The column initially contains a mix of numeric values, missing values (`nan`), and a single non-numeric label `'ACCONCIATORE'`.  

**Goal**: to filter out or modify rows where `zd` contains irrelevant or inconsistent information (e.g., `'ACCONCIATORE'`) to ensure that only meaningful numeric data is retained for analysis.

In [8]:
new_SERVICES.zd.unique()

array(['6', '9', '5', '1', '4', '2', '8', '3', nan, 'ACCONCIATORE', '7'],
      dtype=object)

During the data cleaning process, we identified two consecutive rows in the `SERVICES` dataset that appear to represent the same entity but contain fragmented or overlapping information. Specifically:

1. The first row (index 32) contains descriptive information about the entity, such as `'Acconciatore'` in the `t_es` column, but lacks structured data for most other fields.
2. The second row (index 33) provides structured data for fields like `via`, `civ`, `cod_via`, and `zd`, but the `t_es` column contains a reference to `'ACCONCIATORE'`.

In [9]:
new_SERVICES.loc[32:33].head()

t_es                                  ubic t_via  via   civ  \
32  Acconciatore  CSO COMO N. 15 interno club f. conti   NaN  NaN   NaN   
33      (z.d. 9)                                   CSO  COMO   15  1111   

    cod_via            zd main  sup_alt  sup_lav  
32      NaN           NaN  NaN      NaN      NaN  
33      9.0  ACCONCIATORE  NaN    195.0      NaN

To standardize and improve the dataset:
- The two rows were merged into a single entry, combining the fragmented information into a complete record. 
- This process ensures no information is lost while achieving the desired structure and format for the full dataset.

**Example**:
Before merging:
- Row 32: `'Acconciatore', 'CSO COMO N. 15 interno club f. conti', NaN, NaN, NaN, NaN, NaN, NaN, NaN, NaN`
- Row 33: `'(z.d. 9)', 'CSO', 'COMO', '15', '1111', '9.0', 'ACCONCIATORE', NaN, '195.0', NaN`

After merging:
- New Row: `'Acconciatore', 'CSO COMO', '15', '1111', '9.0', NaN, NaN, '195.0', Nan`

In [10]:
# New merged row
merged_row = {
    "t_es": "Acconciatore",
    "ubic": "CSO COMO N. 15 (z.d. 9)",
    "t_via": "CSO",
    "via": "CSO COMO",
    "civ": "15",
    "cod_via": 1111.0,
    "zd": "9",
    "main": "ACCONCIATORE",
    "sup_alt": np.nan,
    "sup_lav": 195.0,
}

# Remove row at index 33 and replace row at index 32
new_SERVICES = new_SERVICES.drop(index=33, errors='ignore')  # Drop row 33
new_SERVICES.loc[32] = merged_row  # Add the merged row at index 32

# Ensure correct row order and reset index if necessary
new_SERVICES = new_SERVICES.sort_index().reset_index(drop=True)

new_SERVICES.loc[32:33].head()

t_es                                 ubic t_via             via  \
32  Acconciatore              CSO COMO N. 15 (z.d. 9)   CSO        CSO COMO   
33  Acconciatore  CSO DI PORTA NUOVA N. 46 ; (z.d. 1)   CSO  DI PORTA NUOVA   

   civ  cod_via zd          main  sup_alt  sup_lav  
32  15   1111.0  9  ACCONCIATORE      NaN    195.0  
33  46   1044.0  1           NaN      NaN     34.0

Finally, the available values in the column `zd` are:

In [11]:
np.sort(new_SERVICES['zd'].unique())

array(['1', '2', '3', '4', '5', '6', '7', '8', '9'], dtype=object)

## T_ES: Label reduction

The `t_es` column in the dataset contains a variety of labels that represent different categories of services. Many of these labels have similar meanings or represent variations of the same concept, leading to redundancy and potential inconsistencies in analysis.

To address this, we will reduce the number of labels by grouping similar ones under a unified label, making the dataset easier to interpret and analyze while preserving the essential distinctions between categories.

In [12]:
print("The number of different label combination for 't_es' column is {}.".format(new_SERVICES.t_es.unique().shape[0]))

The number of different label combination for 't_es' column is 103.


In [13]:
# Ensure 't_es' column does not have NaN values and is treated as strings
new_SERVICES["t_es"] = new_SERVICES["t_es"].fillna("").astype(str)

# Split the 't_es' column by ';' and flatten the resulting lists
flattened_labels = [label.strip() for item in new_SERVICES["t_es"] for label in item.split(";")]

# Get the unique labels and sort them
unique_labels = sorted(set(flattened_labels))

# Initialize a dictionary to store counts
label_counts = {label: 0 for label in unique_labels}

# Loop through each row in the "t_es" column and count the occurrences of each label
for entry in new_SERVICES['t_es'].dropna():  # Drop NaN values to avoid errors
    for label in unique_labels:
        if label in entry:
            label_counts[label] += 1

print("COUNTS of unique labels\n")

# Find the maximum width for labels and numbers
max_label_length = max(len(label) for label in label_counts if label != "")
max_number_length = max(len(str(count)) for count in label_counts.values())

# Print the counts with aligned output
for label, count in label_counts.items():
    if label != "":
        print(f"{label:<{max_label_length}} \t{count:>{max_number_length}}")

COUNTS of unique labels

ACCONCIATORE                      	 717
Acconciatore                      	 190
BARBIERE                          	   2
Centro abbronzatura               	   2
Centro benessere                  	  29
Centro massaggi                   	 139
Esecuzione di tatuaggi e piercing 	   5
Estetista                         	  31
Estetista in profumeria           	   2
Manicure                          	   3
Parrucchiere misto                	 105
Parrucchiere per signora          	1213
Parrucchiere per uomo             	 518
Pedicure estetico                 	  22
TIPO A - REG.2003                 	 865
TIPO A ESTETICA MANUALE           	 263
TIPO A-B-C-D                      	  88
TIPO B CENTRO DI ABBRONZATURA     	 603
TIPO C TRATT.ESTETICI DIMAGRIM    	 138
TIPO D ESTET.APPAR.ELETTROMECC    	  74
Truccatore                        	   1
esecuzione di tatuaggi e piercing 	  19


Now we transform and standardize the values in the `t_es` column of the dataset. It consists of two main steps:

1. **Composite Labels Handling**: Some rows contain multiple related labels that should be split or expanded into individual categories. For example, labels like `"TIPO A-B-C-D"` or `"Parrucchiere misto"` are expanded into separate labels to ensure all relevant categories are included.
   
2. **Label Mapping**: After handling the composite labels, the code replaces each original label (such as `"Centro abbronzatura"`) with a standardized, consolidated label (e.g., `"Tipo B - Centro di Abbronzatura"`). This step ensures that similar labels are grouped under a single consistent label.

**Example**:

Before transformation:
- `"TIPO A-B-C-D; Parrucchiere misto"`

After transformation:
- `"Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura; Tipo C - Trattamenti Estetici Dimagrimento; Tipo D - Estetica Apparati Elettromeccanici; Parrucchiere per Uomo; Parrucchiere per Donna"`

This process ensures that the dataset contains consistent and meaningful labels, making the data easier to analyze and reducing redundancy.

In [14]:
# Mapping of original values to new values
label_mapping = {
    "ACCONCIATORE": "Acconciatore",
    "Acconciatore": "Acconciatore",
    "BARBIERE": "Parrucchiere per Uomo",
    "Centro abbronzatura": "Tipo B - Centro di Abbronzatura",
    "Centro benessere": "Centro Benessere",
    "Centro massaggi": "Centro Massaggi",
    "Esecuzione di tatuaggi e piercing": "Esecuzione di Tatuaggi e Piercing",
    "esecuzione di tatuaggi e piercing": "Esecuzione di Tatuaggi e Piercing",
    "Estetista": "Tipo A - Estetica Manuale",
    "Estetista in profumeria": "Tipo A - Estetica Manuale",
    "Manicure": "Tipo A - Estetica Manuale", # Assuming merge because there are just 4 'Manicure' values
    "Pedicure estetico": "Tipo A - Estetica Manuale",
    "Parrucchiere per signora": "Parrucchiere per Donna",
    "Parrucchiere per uomo": "Parrucchiere per Uomo",
    "Truccatore": "Truccatore",
    "TIPO A ESTETICA MANUALE": "Tipo A - Estetica Manuale",
    "TIPO A - REG.2003": "Tipo A - Estetica Manuale",
    "TIPO B CENTRO DI ABBRONZATURA": "Tipo B - Centro di Abbronzatura",
    "TIPO C TRATT.ESTETICI DIMAGRIM": "Tipo C - Trattamenti Estetici Dimagrimento",
    "TIPO D ESTET.APPAR.ELETTROMECC": "Tipo D - Estetica Apparati Elettromeccanici",
}

# Define a function for handling composite labels within a cell
def handle_composite_labels(value):
    # If "TIPO A-B-C-D" exists, split and expand into individual components
    if "TIPO A-B-C-D" in value:
        value = value.replace("TIPO A-B-C-D", "; ".join([
            "Tipo A - Estetica Manuale", 
            "Tipo B - Centro di Abbronzatura",
            "Tipo C - Trattamenti Estetici Dimagrimento", 
            "Tipo D - Estetica Apparati Elettromeccanici"
        ]))
    if "Parrucchiere misto" in value:
        value = value.replace("Parrucchiere misto", "; ".join([
            "Parrucchiere per Uomo",
            "Parrucchiere per Donna"
        ]))
    
    return value

# Apply composite label handling first
new_SERVICES["t_es"] = new_SERVICES["t_es"].apply(handle_composite_labels)

# Define a function to replace labels using the mapping
def replace_multiple_labels(cell_value, label_mapping):
    # Split the cell value by a delimiter (semicolon)
    labels = cell_value.split(';')  # Adjust delimiter as necessary
    # Strip whitespace and replace each label using the mapping
    labels = [label_mapping.get(label.strip(), label.strip()) for label in labels]
    
    # Remove duplicates by converting to a set and then back to a list
    labels = list(set(labels))  # Set ensures uniqueness, list keeps order
    # Rejoin the labels with the same delimiter
    return '; '.join(labels)

# Now apply the mapping for individual labels
new_SERVICES["t_es"] = new_SERVICES["t_es"].apply(lambda x: replace_multiple_labels(x, label_mapping))

In [15]:
# Ensure 't_es' column does not have NaN values and is treated as strings
new_SERVICES["t_es"] = new_SERVICES["t_es"].fillna("").astype(str)

# Split the 't_es' column by ';' and flatten the resulting lists
flattened_labels = [label.strip() for item in new_SERVICES["t_es"] for label in item.split(";")]

# Get the unique labels and sort them
unique_labels = sorted(set(flattened_labels))

# Initialize a dictionary to store counts
label_counts = {label: 0 for label in unique_labels}

# Loop through each row in the "t_es" column and count the occurrences of each label
for entry in new_SERVICES['t_es']:  # No need to drop NaN values as we've filled them earlier
    for label in unique_labels:
        if label in entry:
            label_counts[label] += 1

# Find the maximum width for labels and numbers
max_label_length = max(len(label) for label in label_counts if label.strip() != "")
max_number_length = max(len(str(count)) for count in label_counts.values())

print("COUNTS of new unique labels\n")
# Print the counts with aligned output
for label, count in label_counts.items():
    if label.strip() != "":  # Ensure labels are not empty
        print(f"{label:<{max_label_length}} \t{count:>{max_number_length}}")

COUNTS of new unique labels

Acconciatore                                	 907
Centro Benessere                            	  29
Centro Massaggi                             	 139
Esecuzione di Tatuaggi e Piercing           	  24
Parrucchiere per Donna                      	1294
Parrucchiere per Uomo                       	 601
Tipo A - Estetica Manuale                   	1266
Tipo B - Centro di Abbronzatura             	 693
Tipo C - Trattamenti Estetici Dimagrimento  	 226
Tipo D - Estetica Apparati Elettromeccanici 	 162
Truccatore                                  	   1


## MAIN: Label Reduction

In this step, we focus on standardizing the labels in the `main` column. Many of the labels in the column are variations of the same service but are written differently. To ensure consistency across the dataset, we map these different variations to a single, standardized label.

For example, variations like `"CENTRO MASSAGGI"` and `"CENTRO MASSAGGI RILASSANTI NON ESTETICI"` are all mapped to the single label `"Centro Massaggi"`. Similarly, terms like `"ESTETICA A E SOLARIUM"` are transformed into a combination of labels, such as `"Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura"`, to better reflect the service offerings.

This standardization ensures that similar services are grouped together under a consistent set of labels, reducing redundancy and making the dataset more uniform for further analysis.

**Example**:

- Before: `"CENTRO MASSAGGI RILASSANTI NON ESTETICI"` → After: `"Centro Massaggi"`
- Before: `"ESTETICA A E SOLARIUM"` → After: `"Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura"`

In [16]:
print("The number of different label combination  for column 'main' is {}.".format(new_SERVICES.main.unique().shape[0]))

The number of different label combination  for column 'main' is 64.


In [17]:
# Ensure 't_es' column does not have NaN values and is treated as strings
new_SERVICES["main"] = new_SERVICES["main"].fillna("").astype(str)

mapping = {
    'CENTRO MASSAGGI RILASSANTI NON ESTETICI': 'Centro Massaggi',
    'CENTRO MASSAGGI': 'Centro Massaggi',
    'MASSAGGI RILASSANTI': 'Centro Massaggi',
    'CENTRO MASSAGGI TUINA': 'Centro Massaggi',
    'CENTRO MASSAGGI RILASSANTI': 'Centro Massaggi',
    'CENTRO MASSAGI RILASSANTI NON ESTETICI': 'Centro Massaggi', 
    "ATTIVITA' DI MASSAGGI RILASSANTI": 'Centro Massaggi',
    'MASSAGGI': 'Centro Massaggi',
    'centro massaggi': 'Centro Massaggi',
    'CENTRO MASSAGGI THAI RILASSANTI NON CURATIVI': 'Centro Massaggi',
    'CENTRO MASSAGGI RILASSANTI ESTETICI': 'Tipo A - Estetica Manuale; Centro Massaggi',
    'CENTRO BENESSERE': 'Centro Benessere',
    'centro benessere': 'Centro Benessere',
    'CENTRO PER IL BENESSERE FISICO': 'Centro Benessere',
    'CENTTRO BENESSERE (ESCLUSI GLI STABILIMENTI TERMALI)': 'Centro Benessere',
    'CENTRO BENESSERE RILASSANTE NON ESTETICO': 'Centro Benessere',
    'SERVIZI DI CENTRI PER IL BENESSERE FISICO': 'Centro Benessere',
    'SERVIZI DI CENTRI PER IL BENESSERE': 'Centro Benessere',
    'SERVIZI DEI CENTRI E STABIL. PER IL BENESSERE FISICO': 'Centro Benessere',
    'STUDIO OLISTICO': 'Centro Massaggi',
    'CENTRO OLISTICO': 'Centro Massaggi',
    'SERVIZI ORGANIZZATI CON SERVIZI DI SEGRETERIA': 'Tipo A - Estetica Manuale; Centro Massaggi',
    'TRUCCO SEMIPERMANENTE': 'Truccatore',
    'TRUCCATORE': 'Truccatore',
    'TATUAGGI E PIERCING': 'Esecuzione di Tatuaggi e Piercing',
    'ESECUZIONE DI TATUAGGI E PIERCING': 'Esecuzione di Tatuaggi e Piercing',
    'TATUAGGIO': 'Esecuzione di Tatuaggi e Piercing',
    'TATUAGGIO E PIERCING': 'Esecuzione di Tatuaggi e Piercing',
    'ESECUZIONE DI TATUAGGI': 'Esecuzione di Tatuaggi e Piercing',
    'ESECUZIONE TATUAGGI': 'Esecuzione di Tatuaggi e Piercing',
    'tatuaggi e piercing': 'Esecuzione di Tatuaggi e Piercing',
    'ACCONCIATORE': 'Acconciatore',
    'acconciatore': 'Acconciatore',
    'acconciatore.': 'Acconciatore',
    'ACONCIATORE': 'Acconciatore',
    'ACCONCIATORE UOMO DONNA': 'Acconciatore',
    'ACCONCIATORE E ESTETICA': 'Acconciatore; Tipo A - Estetica Manuale',
    'SERVIZI DEI SALONI DI BARBIERE E PARRUCCHIERE': 'Acconciatore',
    'PARRUCCHIERE UOMO DONNA BARBIERE': 'Acconciatore',
    'PARRUCCHIERE': 'Acconciatore',
    'ESTETISTA': 'Tipo A - Estetica Manuale',
    'ESTETICA': 'Tipo A - Estetica Manuale',
    'ESTETICA TIPO A': 'Tipo A - Estetica Manuale',
    'estetica tipo A': 'Tipo A - Estetica Manuale',
    'ESTETISTA MANICURE PEDICURE': 'Tipo A - Estetica Manuale',
    'estetista': 'Tipo A - Estetica Manuale',
    'estetica': 'Tipo A - Estetica Manuale',
    'APPLICAZIONE E DECORAZIONE UNGHIE ARTIFICIALE': 'Tipo A - Estetica Manuale',
    'RICOSTRUZIONE UNGHIE': 'Tipo A - Estetica Manuale',
    'ESTETICA - CENTRO BENESSERE': 'Tipo A - Estetica Manuale; Centro Benessere',
    'ESTETICA A': 'Tipo A - Estetica Manuale',
    'estetica A': 'Tipo A - Estetica Manuale',
    'ESTETICA a': 'Tipo A - Estetica Manuale',
    'ESTETICA A E SOLARIUM': 'Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura',
    'solarium': 'Tipo B - Centro di Abbronzatura', 
    'ESTETICA A CON TATUAGGI': 'Tipo A - Estetica Manuale; Esecuzione di Tatuaggi e Piercing',
    'ESTETICA A e B': 'Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura',
    'ESTETICA A B': 'Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura',
    'ESTETICA A E B': 'Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura',
    'estetica A e solarium': 'Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura',
    'SOLARIUM': 'Tipo B - Centro di Abbronzatura',
    'SOLARIUM E ESTETICA DI TIPO A': 'Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura',
    'ESTETICA TIPO A - B': 'Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura',
    'CENTRO ESTETICO - SOLARIUN': 'Tipo A - Estetica Manuale; Tipo B - Centro di Abbronzatura'
}

def replace_multiple_labels(cell_value, label_mapping):
    # Split the cell value by a delimiter (semicolon)
    labels = cell_value.split(';')  # Adjust delimiter as necessary
    # Strip whitespace and replace each label using the mapping
    labels = [label_mapping.get(label.strip(), label.strip()) for label in labels]
    # Join the labels back with the same delimiter
    return ';'.join(labels)

# Now apply the mapping for individual labels
new_SERVICES["main"] = new_SERVICES["main"].apply(lambda x: replace_multiple_labels(x, mapping))

In [18]:
# Initialize a dictionary to store counts
flattened_labels = [label.strip() for item in new_SERVICES["main"] for label in item.split(";")]
# Get the unique labels and sort them
unique_main_labels = sorted(set(flattened_labels))

label_counts = {label: 0 for label in unique_main_labels}

# Loop through each row in the "main" column and count the occurrences of each label
for entry in new_SERVICES['main'].dropna():  # Drop NaN values to avoid errors
    for label in unique_main_labels:
        if label in entry:
            label_counts[label] += 1

# Find the maximum width for labels and numbers
max_label_length = max(len(label) for label in label_counts if label != "")
max_number_length = max(len(str(count)) for count in label_counts.values())

print("COUNTS of new unique labels\n")
# Print the counts with aligned output
for label, count in label_counts.items():
    if label != "":
        print(f"{label:<{max_label_length}} \t{count:>{max_number_length}}")

COUNTS of new unique labels

Acconciatore                      	 100
Centro Benessere                  	  13
Centro Massaggi                   	  98
Esecuzione di Tatuaggi e Piercing 	  19
Tipo A - Estetica Manuale         	  66
Tipo B - Centro di Abbronzatura   	  20
Truccatore                        	   1


Once standardized the column, we decide to move the values from the `main` column to the `t_es` column. 

The `main` column contains many null values, and the few non-null values it holds are not significantly different from the information already present in the `t_es` column. Therefore, we use the `main` column values to populate the `t_es` column where `t_es` is null, ensuring that we maintain all relevant information in a single column.

After this step, the `main` column will be dropped, as its contribution is minimal and redundant.

In [19]:
new_SERVICES['t_es'] = new_SERVICES['t_es'].replace("", float("nan"))
print("The initial number of Nan in 't_es' column is {}.".format(sum(new_SERVICES["t_es"].isnull())))

The initial number of Nan in 't_es' column is 31.


In [20]:
# Ensure both columns are strings and handle NaN values
new_SERVICES['main'] = new_SERVICES['main'].fillna("").astype(str)
new_SERVICES['t_es'] = new_SERVICES['t_es'].fillna("").astype(str)

def update_t_es(row):
    main_values = row['main'].split(";") if row['main'] else []
    t_es_values = row['t_es'].split(";") if row['t_es'] else []
    
    # Remove extra whitespace around the labels
    main_values = [value.strip() for value in main_values]
    t_es_values = [value.strip() for value in t_es_values]
    
    # If t_es is empty, copy main values
    if not t_es_values:
        return ";".join(main_values)
    
    # Add main labels that are not already in t_es
    for main_value in main_values:
        # Add the condition: If main is "Acconciatore" and t_es contains "Parrucchiere per Uomo" or "Parrucchiere per Donna", don't add
        if main_value == "Acconciatore" and any(x in t_es_values for x in ["Parrucchiere per Uomo", "Parrucchiere per Donna"]):
            continue
        if main_value not in t_es_values:
            t_es_values.append(main_value)
    
    # Return the updated t_es column as a semicolon-separated string
    return ";".join(t_es_values)

# Apply the function row-wise
new_SERVICES['t_es'] = new_SERVICES.apply(update_t_es, axis=1)

In [21]:
new_SERVICES['t_es'] = new_SERVICES['t_es'].replace("", float("nan"))
print("The number of Nan in 't_es' column afterwards is {}.".format(sum(new_SERVICES["t_es"].isnull())))

The number of Nan in 't_es' column afterwards is 16.


Since `main` info is incorporated in `t_es` column, we drop it.

In [22]:
new_SERVICES = new_SERVICES.drop('main', axis=1)

new_SERVICES.head()

t_es                                 ubic t_via  \
0               NaN     LGO DEI GELSOMINI N. 10 (z.d. 6)   LGO   
1   Centro Massaggi              PZA FIDIA N. 3 (z.d. 9)   PZA   
2  Centro Benessere             VIA ADIGE N. 10 (z.d. 5)   VIA   
3        Truccatore  VIA BARACCHINI FLAVIO N. 9 (z.d. 1)   VIA   
4               NaN           VIA BERGAMO N. 12 (z.d. 4)   VIA   

                 via civ  cod_via zd  sup_alt  sup_lav  
0      DEI GELSOMINI  10   5394.0  6      NaN     55.0  
1              FIDIA   3   1144.0  9      2.0     28.0  
2              ADIGE  10   4216.0  5      2.0     27.0  
3  BARACCHINI FLAVIO   9    356.0  1      NaN      NaN  
4            BERGAMO  12   3189.0  4      NaN     50.0

## T_VIA: Checking format
In this phase, we verify that the `t_via` column contains only valid and properly formatted values.

The column contains designations for the type of street or location, which are standardized according to the tipovia dataframe. This dataframe defines valid labels and their corresponding extended and abbreviated descriptions.

In [23]:
# Ensure comparison is case-insensitive by converting both columns to lowercase
t_via_unique = new_SERVICES['t_via'].unique()
desc_abbreviata = tipovia['DESC_ABBREVIATA'].str.lower().unique()

# Find t_via values not in tipovia.DESC_ABBREVIATA
invalid_t_via = [via for via in t_via_unique if via.lower() not in desc_abbreviata]

# Print the results
print("Invalid t_via values not in tipovia:\n")
print(invalid_t_via)


Invalid t_via values not in tipovia:

['VIE']


Since any potential unformatted or invalid values in the `t_via` column were already addressed during the previous merging and correction phases (e.g., handling rows with `"Acconciatore"`), only one inadmissible value remains: `"VIE"`. This value is not part of the valid labels defined in the `DESC_ABBREVIATA` column of the **tipovia** dataframe.

To resolve this, the `"VIE"` value must be corrected or mapped to a valid designation from the **tipovia** dataframe.

In [24]:
new_SERVICES[new_SERVICES["t_via"] == "VIE"]

t_es  \
721                                                                      Acconciatore   
2891  Tipo B - Centro di Abbronzatura;Tipo A - Estetica Manuale;Parrucchiere per Uomo   

                                       ubic t_via                via civ  \
721   VIE DELLA ROBBIA LUCA N. 8 ; (z.d. 8)   VIE  DELLA ROBBIA LUCA   8   
2891  VIE DELLA ROBBIA LUCA N. 8 ; (z.d. 8)   VIE  DELLA ROBBIA LUCA   8   

      cod_via zd  sup_alt  sup_lav  
721    6349.0  8      NaN     33.0  
2891   6349.0  8      NaN      NaN

In [25]:
new_SERVICES.loc[new_SERVICES["t_via"] == "VIE", "t_via"] = "VIA"

Once this adjustment is made, we ensure that all `t_via` values match the `DESC_ABBREVIATA` column of **tipovia** and can be directly mapped to their corresponding `DESC_ESTESA` values.

In [26]:
# Create a mapping dictionary from tipovia
t_via_mapping = dict(zip(tipovia["DESC_ABBREVIATA"].str.lower(), tipovia["DESC_ESTESA"]))

# Replace the t_via values in new_SERVICES with the corresponding DESC_ESTESA values
new_SERVICES["t_via"] = new_SERVICES["t_via"].str.lower().map(t_via_mapping)

# Check the result
new_SERVICES.head()

t_es                                 ubic   t_via  \
0               NaN     LGO DEI GELSOMINI N. 10 (z.d. 6)   Largo   
1   Centro Massaggi              PZA FIDIA N. 3 (z.d. 9)  Piazza   
2  Centro Benessere             VIA ADIGE N. 10 (z.d. 5)     Via   
3        Truccatore  VIA BARACCHINI FLAVIO N. 9 (z.d. 1)     Via   
4               NaN           VIA BERGAMO N. 12 (z.d. 4)     Via   

                 via civ  cod_via zd  sup_alt  sup_lav  
0      DEI GELSOMINI  10   5394.0  6      NaN     55.0  
1              FIDIA   3   1144.0  9      2.0     28.0  
2              ADIGE  10   4216.0  5      2.0     27.0  
3  BARACCHINI FLAVIO   9    356.0  1      NaN      NaN  
4            BERGAMO  12   3189.0  4      NaN     50.0

## COD_VIA: Checking Format
This check ensures that the values in the `cod_via` column, which are stored as floats, are whole numbers (i.e., integers), since the identifiers related to street or location codes should always be integers.

In [27]:
# Check if any value in 'cod_via' is not an integer (checking for float values that are whole numbers)
non_integer_values = new_SERVICES['cod_via'].apply(lambda x: not x.is_integer() if isinstance(x, float) else False)

if non_integer_values.any():
    print("There are values in 'cod_via' that are not integers.")
else:
    print("All values in 'cod_via' are integers (float equivalents of integers).")

new_SERVICES.cod_via = new_SERVICES.cod_via.astype(float).astype(int)

All values in 'cod_via' are integers (float equivalents of integers).


## CIV: Checking Format
This code checks the `civ` column to ensure that all values follow the expected format: **1-3 digit, optionally followed by an uppercase letter (e.g., 12A), a slash "/" with a single digit (e.g., 93/1), or a slash "/" and another series of digits and letters (e.g., 15/101).

In [28]:
# Define a function to check the format
def check_civ_format(value):
    # Regex pattern to match 1-3 digits, optionally followed by a `/` and a single digit or an uppercase letter
    pattern = r'^\d{1,3}(?:/\d{1,3}[A-Z]?)?(?:[A-Z]+)?(?:[A-Z]+P\d{2})?(?:[A-Z]+N\d{2})?(?:[A-Z]+C\d{2})?(?:[A-Z]+G\d{2})?$'
    return bool(re.match(pattern, str(value)))

# Apply the check to the 'civ' column and filter invalid values
invalid_civ_values = new_SERVICES['civ'][new_SERVICES['civ'].apply(lambda x: not check_civ_format(x) if pd.notnull(x) else False)]

if not invalid_civ_values.empty:
    print("There are invalid values in 'civ' column.\n")
    
if not invalid_civ_values.empty:
    for idx, value in invalid_civ_values.items():
        print(f"The row with index {idx} contains an invalid value: {value}.")
else:
    print("All values in 'civ' follow the correct format.")

There are invalid values in 'civ' column.

The row with index 143 contains an invalid value: 40945.


We handle invalid values in the `civ` column by replacing them with `NaN`, ensuring that only correctly formatted entries remain for further processing.

In [29]:
new_SERVICES[143:].head(1)

t_es                       ubic t_via    via    civ  cod_via zd  \
143  Acconciatore  VIA PAVIA N. 6/2 (z.d. 5)   Via  PAVIA  40945     5262  5   

     sup_alt  sup_lav  
143      NaN      NaN

In [30]:
# Replace invalid 'civ' values with NaN
new_SERVICES['civ'] = new_SERVICES['civ'].apply(lambda x: np.nan if not check_civ_format(x) else x)
new_SERVICES[143:].head(1)

t_es                       ubic t_via    via  civ  cod_via zd  \
143  Acconciatore  VIA PAVIA N. 6/2 (z.d. 5)   Via  PAVIA  NaN     5262  5   

     sup_alt  sup_lav  
143      NaN      NaN

In [31]:
print("The number of different 'civ' values for column 'main' is {}.".format(new_SERVICES.civ.unique().shape[0]))

The number of different 'civ' values for column 'main' is 234.


## VIA: Removal of Redundant Information and Validation

When processing records, some entries contain redundant information in the `via` column that matches part of the `t_via` column.
To ensure consistency and follow a specific order, we remove the redundant portion.

**Example**

Given the following data:

| t_es          | ubic                      | t_via | via      | civ | cod_via | zd | sup_alt | sup_lav |
|---------------|---------------------------|-------|----------|-----|---------|----|---------|---------|
| Acconciatore  | CSO COMO N. 15 (z.d. 9)  | CSO   | CSO COMO     | 15  | 1111.0  | 9  | NaN     | 195.0   |


The updated record becomes:

| t_es          | ubic                      | t_via | via      | civ | cod_via | zd | sup_alt | sup_lav |
|---------------|---------------------------|-------|----------|-----|---------|----|---------|---------|
| Acconciatore  | CSO COMO N. 15 (z.d. 9)  | CSO   | COMO     | 15  | 1111.0  | 9  | NaN     | 195.0   |


In this transformation:
- The `via` column value `CSO COMO` had the redundant prefix `CSO` matching the value in the `t_via` column.
- The prefix was removed, leaving only the unique portion: `COMO`.

In [32]:
new_SERVICES.via = new_SERVICES.via.fillna("").astype(str)
new_SERVICES.t_via = new_SERVICES.t_via.fillna("").astype(str) 

# Function to remove the redundant prefix
def clean_via(row):
    if row["via"].startswith(row["t_via"] + " ") and row["t_via"] != "":  # Check if via starts with t_via
        return row["via"].replace(row["t_via"] + " ", "")  # Remove t_via + space
    return row["via"]  # Return original if no match

# Apply the function to the 'via' column
new_SERVICES["via"] = new_SERVICES.apply(clean_via, axis=1)

new_SERVICES[new_SERVICES.via == "COMO"].head(1)

t_es                       ubic  \
1140  Parrucchiere per Donna;Parrucchiere per Uomo  CSO COMO N. 11 ; (z.d. 9)   

      t_via   via civ  cod_via zd  sup_alt  sup_lav  
1140  Corso  COMO  11     1111  9      NaN     23.0

Now we validate the `via` column in the dataset by comparing it against the descriptive fields (`DESCRITTIVO` and `DENOMINAZIONE`) from the **viario** dataframe.<br> We first normalize the text by converting it to lowercase, removing unnecessary spaces and parentheses, and sorting the words alphabetically. Then, we calculate similarity scores between the `via` column and the two descriptive fields, using the highest score to flag rows for review if they fall below a **75%** similarity threshold. 

In [33]:
# Function to normalize the names by removing extra spaces, converting to lowercase, and sorting words
def normalize_string(s):

    #to remove specific pattern contained in the data
    s = re.sub(r"sant'\s*", "sant'", s.lower())  # Fix spaces around 'sant'
    s = re.sub(r"dell'\s*", "dell'", s.lower())  # Fix spaces around 'dell
    s = re.sub(r"[()]", "", s) # Remove the parenthesis
    
    # Split the string into words
    words = s.strip().split()
    
    # Sort the words alphabetically
    words.sort()
    
    # Join the words back together and convert to lowercase
    return " ".join(words)

# Function to compute the edit distance between two strings (after normalization)
def calculate_similarity(str1, str2):
    return fuzz.ratio(str1, str2)

# Merge new_SERVICES with viario on 'cod_via' and 'CODICE_VIA' to bring in 'DESCRITTIVO'
merged_df = new_SERVICES.merge(viario[['CODICE_VIA', 'DESCRITTIVO', 'DENOMINAZIONE']], left_on='cod_via', right_on='CODICE_VIA', how='left')

# Normalize 'via' and 'DESCRITTIVO' columns
merged_df['normalized_via'] = merged_df['via'].apply(normalize_string)
merged_df['normalized_descrittivo'] = merged_df['DESCRITTIVO'].apply(normalize_string)
merged_df['normalized_denominazione'] = merged_df['DENOMINAZIONE'].apply(normalize_string)

# Calculate the similarity between normalized values using edit distance
merged_df['similarity'] = merged_df.apply(lambda row: max(calculate_similarity(row['normalized_via'], row['normalized_descrittivo']), calculate_similarity(row['normalized_via'], row['normalized_denominazione'])), axis=1)

# Print rows where the similarity is below a certain threshold (e.g., 80)
threshold = 75
non_matching_rows = merged_df[merged_df['similarity'] < threshold]

# Show the rows with low similarity
print(f"Rows where normalized 'via' and 'DESCRITTIVO'/'DENOMINAZIONE' are not similar enough (below {threshold}% similarity):")
non_matching_rows[['cod_via', 'via', 'DENOMINAZIONE', 'DESCRITTIVO', 'similarity']].sort_values(by="similarity", ascending=True)

Rows where normalized 'via' and 'DESCRITTIVO'/'DENOMINAZIONE' are not similar enough (below 75% similarity):


cod_via                    via                  DENOMINAZIONE  \
3087      612        DANTE ALIGHIERI                          DANTE   
2558     1524    GRAZIANO IMPERATORE                       GRAZIANO   
3322     1524    GRAZIANO IMPERATORE                       GRAZIANO   
2308     6380       TIZIANO VECELLIO                       TIZIANO    
3453     6380       TIZIANO VECELLIO                       TIZIANO    
363      7158  COLONNA MARCO ANTONIO            COLONNA MARCANTONIO   
2868     7158  COLONNA MARCO ANTONIO            COLONNA MARCANTONIO   
3612     7158  COLONNA MARCO ANTONIO            COLONNA MARCANTONIO   
32       1111               CSO COMO                           COMO   
1572     4144              DON BOSCO             BOSCO DON GIOVANNI   
1573     4144              DON BOSCO             BOSCO DON GIOVANNI   
3298     4144              DON BOSCO             BOSCO DON GIOVANNI   
3631     4144              DON BOSCO             BOSCO DON GIOVANNI   
2223     6551           MURILLO S.B.  MURILLO SEBASTIANO BARTOLOMEO   
3490     6551           MURILLO S.B.  MURILLO SEBASTIANO BARTOLOMEO   

              DESCRITTIVO  similarity  
3087                DANTE   50.000000  
2558             GRAZIANO   59.259259  
3322             GRAZIANO   59.259259  
2308              TIZIANO   60.869565  
3453              TIZIANO   60.869565  
363   MARCANTONIO COLONNA   65.000000  
2868  MARCANTONIO COLONNA   65.000000  
3612  MARCANTONIO COLONNA   65.000000  
32                   COMO   66.666667  
1572   DON GIOVANNI BOSCO   66.666667  
1573   DON GIOVANNI BOSCO   66.666667  
3298   DON GIOVANNI BOSCO   66.666667  
3631   DON GIOVANNI BOSCO   66.666667  
2223              MURILLO   73.684211  
3490              MURILLO   73.684211

Since these values with similarity lower than 75% match the respective addresses semantically, the `via` column can now be standardized to match the format of the `viario.DESCRITTIVO` column.

In [34]:
# Merge new_SERVICES with viario on the matching cod_via and CODICE_VIA
new_SERVICES = new_SERVICES.merge(viario[['CODICE_VIA', 'DESCRITTIVO']], 
                                  left_on='cod_via', right_on='CODICE_VIA', 
                                  how='left')

# Update the via column with the corresponding DESCRITTIVO value
new_SERVICES['via'] = new_SERVICES['DESCRITTIVO']

# Drop the extra columns (if needed)
new_SERVICES.drop(columns=['CODICE_VIA', 'DESCRITTIVO'], inplace=True)
new_SERVICES.head()

t_es                                 ubic   t_via  \
0               NaN     LGO DEI GELSOMINI N. 10 (z.d. 6)   Largo   
1   Centro Massaggi              PZA FIDIA N. 3 (z.d. 9)  Piazza   
2  Centro Benessere             VIA ADIGE N. 10 (z.d. 5)     Via   
3        Truccatore  VIA BARACCHINI FLAVIO N. 9 (z.d. 1)     Via   
4               NaN           VIA BERGAMO N. 12 (z.d. 4)     Via   

                 via civ  cod_via zd  sup_alt  sup_lav  
0      DEI GELSOMINI  10     5394  6      NaN     55.0  
1              FIDIA   3     1144  9      2.0     28.0  
2              ADIGE  10     4216  5      2.0     27.0  
3  FLAVIO BARACCHINI   9      356  1      NaN      NaN  
4            BERGAMO  12     3189  4      NaN     50.0

## UBIC: Structuring and Format Changing
We process the `ubic` field to extract and organize its contents into distinct components: `ubic_t_via` (street type), `ubic_via` (street name), `ubic_civ` (street number), `ubic_zd` (zone description), and `note` (additional information). This step allows us to compare these extracted values with the corresponding columns that share similar semantic meanings (`t_via`, `via`, `civ`, and `zd`). 

Once the comparison is complete, we remove the `ubic` field and its derived columns to eliminate redundancy. 

This approach ensures a cleaner and more structured dataset for further analysis.

**Example**

Given the following data:

| t_es  | ubic | t_via | via | civ | cod_via | zd | sup_alt | sup_lav | 
|-------|------|-------|-----|-----|---------|----|---------|---------|
| Tipo A - Estetica Manuale;Tipo C - Trattamenti Estetici Dimagrimento | ALZ NAVIGLIO PAVESE N. 8 int.attivita' di parrucchiere; (z.d. 6) | ALZ | NAVIGLIO PAVESE | 8 | 5161.0 | 6 | NaN | NaN |

The updated record becomes:

| t_es  | t_via | via | civ | cod_via | zd | sup_alt | sup_lav | ubic_t_via | ubic_via | ubic_civ | ubic_zd | note |
|-------|-------|-----|-----|---------|----|---------|---------|------------|----------|----------|--------|------|
| Tipo A - Estetica Manuale;Tipo C - Trattamenti Estetici Dimagrimento  | ALZ | NAVIGLIO PAVESE | 8 | 5161.0 | 6 | NaN | NaN | ALZ | NAVIGLIO PAVESE | 8 | 6 | int.attivita' di parrucchiere |

**Explanation**:

In this transformation:
- The `via` column value `NAVIGLIO PAVESE` had the redundant prefix `ALZ` matching the value in the `t_via` column.
- The prefix `ALZ` is removed from the `via` column, leaving the unique portion: `NAVIGLIO PAVESE`.
- The `ubic` field remains the same in structure but is used to extract the relevant information, including the street name, number, and zone description, which are now placed in separate columns (`ubic_via`, `ubic_civ`, `ubic_zd`), while any additional information (e.g., `int.attivita' di parrucchiere`) is placed in the `note` column.

In [35]:
new_SERVICES['ubic'] = new_SERVICES['ubic'].fillna("").astype(str)

def split_ubicazione(ubicazione):
    # Initialize variables
    tipo_via = np.nan
    via = np.nan
    civico = np.nan
    zd = np.nan
    note = np.nan
    
    ubicazione = ubicazione.replace("ø", "A")   # Handle lone 'Ã'
    
    # Split the string into parts
    parts = ubicazione.split(" ")

    if parts[0].lower() == "vie":
        parts[0] = "via"
    
    # Extract "Tipo Via" (first word if valid)
    if parts[0].lower() in tipovia.DESC_ABBREVIATA.str.lower().unique():
        tipo_via = parts[0]
        parts = parts[1:]  # Remove 'Tipo Via' from further processing

    # Find "Civico" (number after "N." or "num.")
    match = re.search(r'\b(?:N\.|num\.)\s*([\w]+)', ubicazione, re.IGNORECASE)
    if match:
        civico = match.group(1).strip(";")
        parts = re.sub(r'\b(?:N\.|num\.)\s*[\w]+', '', ubicazione).split()

    # Extract "ZD" (zone description)
    match_zd = re.search(r'\(z\.d\.\s*(\d+)\)', ubicazione, re.IGNORECASE)
    if match_zd:
        zd = match_zd.group(1)
        parts = re.sub(r'\(z\.d\.\s*\d+\)', '', " ".join(parts)).split()

    # To put together strings of the note like '1A' and 'piano' 
    # Check if there are at least three words to avoid IndexError
    if len(parts) > 3:
        # Define a regular expression to match "number + letter" (e.g., "1A")
        pattern = r"^\d+[A-Za-z]$"
        
        # Check if either parts[-2] or parts[-1] matches the pattern
        if re.match(pattern, parts[-2]) or re.match(pattern, parts[-1]):
            # Concatenate third-to-last and second-to-last parts
            parts[-2] = parts[-2] + " " + parts[-1]
            del parts[-1]  # Remove the last word after merging

    # Identify and extract the "Via" (remaining uppercase parts)
    via_parts = []
    note_parts = []
    for part in parts:
        if part.isupper():  # Uppercase letters are part of "Via"
            via_parts.append(part)
        else:  # Lowercase letters are considered "Note"
            note_parts.append(part)

    # Construct the "Via" without the "Tipo Via"
    via_parts = [part for part in via_parts if part != tipo_via]  # Remove 'tipo_via' from 'via_parts'
    via = " ".join(via_parts).strip()
    via =  via.replace(";", "")
    note = " ".join(note_parts).strip(";") if note_parts else np.nan
    if note == "":
        note = np.nan
    
    # Remove leading zeros unless the value contains alphabetic characters
    if civico and civico.isdigit():
        civico = str(int(civico))  # Remove leading zeros by converting to an integer
    elif civico:  # Preserve alphanumeric strings like '012B'
        civico = civico.lstrip("0").upper()  # Strip leading zeros but keep characters intact

    return tipo_via, via, civico, zd, note

# Apply the function to create new columns
new_SERVICES[['ubic_t_via', 'ubic_via', 'ubic_civ', 'ubic_zd', 'note']] = new_SERVICES['ubic'].apply(lambda x: pd.Series(split_ubicazione(x)))
new_SERVICES.head()

t_es                                 ubic   t_via  \
0               NaN     LGO DEI GELSOMINI N. 10 (z.d. 6)   Largo   
1   Centro Massaggi              PZA FIDIA N. 3 (z.d. 9)  Piazza   
2  Centro Benessere             VIA ADIGE N. 10 (z.d. 5)     Via   
3        Truccatore  VIA BARACCHINI FLAVIO N. 9 (z.d. 1)     Via   
4               NaN           VIA BERGAMO N. 12 (z.d. 4)     Via   

                 via civ  cod_via zd  sup_alt  sup_lav ubic_t_via  \
0      DEI GELSOMINI  10     5394  6      NaN     55.0        LGO   
1              FIDIA   3     1144  9      2.0     28.0        PZA   
2              ADIGE  10     4216  5      2.0     27.0        VIA   
3  FLAVIO BARACCHINI   9      356  1      NaN      NaN        VIA   
4            BERGAMO  12     3189  4      NaN     50.0        VIA   

            ubic_via ubic_civ ubic_zd note  
0      DEI GELSOMINI       10       6  NaN  
1              FIDIA        3       9  NaN  
2              ADIGE       10       5  NaN  
3  BARACCHINI FLAVIO        9       1  NaN  
4            BERGAMO       12       4  NaN

In [36]:
new_SERVICES = new_SERVICES.drop("ubic", axis=1)
new_SERVICES.head()

t_es   t_via                via civ  cod_via zd  sup_alt  \
0               NaN   Largo      DEI GELSOMINI  10     5394  6      NaN   
1   Centro Massaggi  Piazza              FIDIA   3     1144  9      2.0   
2  Centro Benessere     Via              ADIGE  10     4216  5      2.0   
3        Truccatore     Via  FLAVIO BARACCHINI   9      356  1      NaN   
4               NaN     Via            BERGAMO  12     3189  4      NaN   

   sup_lav ubic_t_via           ubic_via ubic_civ ubic_zd note  
0     55.0        LGO      DEI GELSOMINI       10       6  NaN  
1     28.0        PZA              FIDIA        3       9  NaN  
2     27.0        VIA              ADIGE       10       5  NaN  
3      NaN        VIA  BARACCHINI FLAVIO        9       1  NaN  
4     50.0        VIA            BERGAMO       12       4  NaN

Now we standardize the `ubic_t_via` column with the values from the **tipovia** dataframe. Since we have already ensured that all values in `ubic_t_via` match the corresponding abbreviations from the `DESC_ABBREVIATA` column in **tipovia**, there is no need for additional corrections or formatting. We directly map the `ubic_t_via` values to their corresponding `DESC_ESTESA` values from **tipovia** as `desc_attesa`.

In [37]:
# Create a mapping dictionary from tipovia for ubic_t_via
ubic_t_via_mapping = dict(zip(tipovia["DESC_ABBREVIATA"].str.lower(), tipovia["DESC_ESTESA"]))

# Replace the ubic_t_via values in new_SERVICES with the corresponding DESC_ESTESA values
new_SERVICES["ubic_t_via"] = new_SERVICES["ubic_t_via"].str.lower().map(ubic_t_via_mapping)

# Check the result
new_SERVICES.head()

t_es   t_via                via civ  cod_via zd  sup_alt  \
0               NaN   Largo      DEI GELSOMINI  10     5394  6      NaN   
1   Centro Massaggi  Piazza              FIDIA   3     1144  9      2.0   
2  Centro Benessere     Via              ADIGE  10     4216  5      2.0   
3        Truccatore     Via  FLAVIO BARACCHINI   9      356  1      NaN   
4               NaN     Via            BERGAMO  12     3189  4      NaN   

   sup_lav ubic_t_via           ubic_via ubic_civ ubic_zd note  
0     55.0      Largo      DEI GELSOMINI       10       6  NaN  
1     28.0     Piazza              FIDIA        3       9  NaN  
2     27.0        Via              ADIGE       10       5  NaN  
3      NaN        Via  BARACCHINI FLAVIO        9       1  NaN  
4     50.0        Via            BERGAMO       12       4  NaN

Next, we validate the `ubic_via` column in the dataset by comparing it to the descriptive fields (`DESCRITTIVO` and `DENOMINAZIONE`) from the `viario` dataframe, following a similar process as for the `via` column. <br> First, we normalize the text by converting it to lowercase, removing extraneous spaces and parentheses (while preserving their contents), and sorting the words alphabetically to address potential inconsistencies. Then, we calculate similarity scores between the `ubic_via` column and both `DESCRITTIVO` and `DENOMINAZIONE`. The highest similarity score is selected for each row, and rows with a similarity score below 75% are flagged for further review.

In [38]:
# Function to normalize the names
def normalize_string(s):
    s = re.sub(r"sant'\s*", "sant'", s.lower())  # Fix spaces around 'sant'
    s = re.sub(r"dell'\s*", "dell'", s.lower())  # Fix spaces around 'dell'
    s = re.sub(r"[()]", "", s)  # Remove the parentheses
    
    # Split the string into words, sort them, and rejoin
    words = s.strip().split()
    words.sort()
    return " ".join(words)

# Pre-normalize the viario DataFrame
viario['normalized_descrittivo'] = viario['DESCRITTIVO'].apply(normalize_string)
viario['normalized_denominazione'] = viario['DENOMINAZIONE'].apply(normalize_string)

# Function to find the best match using pre-normalized viario
def find_best_match(ubic_via):

    measure = "EDIT"
    normalized_ubic_via = normalize_string(ubic_via)
    
    # Compute similarity scores for all rows in viario
    viario['similarity_descrittivo'] = viario['normalized_descrittivo'].apply(
        lambda x: fuzz.ratio(normalized_ubic_via, x))
    viario['similarity_denominazione'] = viario['normalized_denominazione'].apply(
        lambda x: fuzz.ratio(normalized_ubic_via, x))
    
    # Find the best match
    viario['similarity'] = viario[['similarity_descrittivo', 'similarity_denominazione']].max(axis=1)
    best_match = viario.loc[viario['similarity'].idxmax()]
    
    return (measure, best_match['similarity'],
            ubic_via, best_match['DENOMINAZIONE'],best_match['DESCRITTIVO'],
            normalized_ubic_via, best_match['normalized_denominazione'], best_match['normalized_descrittivo'],
            )

# Process all rows in new_SERVICES
results = []

for ubic_via in tqdm(new_SERVICES['ubic_via'].unique()):
    ubic_via_data = find_best_match(ubic_via)
    results.append(ubic_via_data)

# Convert results to DataFrame for analysis
columns = ['measure','similarity',
            'ubic_via', 'best_denominazione', 'best_descrittivo', 'normalized_ubic_via',
            'best_normalized_denominazione', 'best_normalized_descrittivo', 
            ]
check_matching_df = pd.DataFrame(results, columns=columns)

# Show the rows with low similarity
print(f"Rows where normalized 'ubic_via' and 'DESCRITTIVO'/'DENOMINAZIONE' are not similar enough (below {threshold}% Edit Distance similarity):")
check_matching_df.sort_values(by="similarity", ascending=True)

100%|██████████| 1381/1381 [00:09<00:00, 143.53it/s]

Rows where normalized 'ubic_via' and 'DESCRITTIVO'/'DENOMINAZIONE' are not similar enough (below 75% Edit Distance similarity):


measure  similarity                 ubic_via       best_denominazione  \
183     EDIT    0.000000                                       DUOMO (DEL)   
1172    EDIT   63.157895      GRAZIANO IMPERATORE      VIMERCATI OTTAVIANO   
1103    EDIT   64.705882         TIZIANO VECELLIO       AMMIANO MARCELLINO   
1265    EDIT   68.750000          DANTE ALIGHIERI        CHIASSERINI DANTE   
292     EDIT   71.428571    COLONNA MARCO ANTONIO    CARLONE CARLO ANTONIO   
...      ...         ...                      ...                      ...   
461     EDIT  100.000000  VENINI GIULIO E CORRADO  VENINI GIULIO E CORRADO   
460     EDIT  100.000000                   VEGLIA                   VEGLIA   
459     EDIT  100.000000                   VARESE                   VARESE   
467     EDIT  100.000000                 VIMINALE                 VIMINALE   
1380    EDIT  100.000000                   NIRONE                   NIRONE   

             best_descrittivo      normalized_ubic_via  \
183                 DEL DUOMO                            
1172      OTTAVIANO VIMERCATI      graziano imperatore   
1103       MARCELLINO AMMIANO         tiziano vecellio   
1265        DANTE CHIASSERINI          alighieri dante   
292     CARLO ANTONIO CARLONE    antonio colonna marco   
...                       ...                      ...   
461   GIULIO E CORRADO VENINI  corrado e giulio venini   
460                    VEGLIA                   veglia   
459                    VARESE                   varese   
467                  VIMINALE                 viminale   
1380                   NIRONE                   nirone   

     best_normalized_denominazione best_normalized_descrittivo  
183                      del duomo                   del duomo  
1172           ottaviano vimercati         ottaviano vimercati  
1103            ammiano marcellino          ammiano marcellino  
1265             chiasserini dante           chiasserini dante  
292          antonio carlo carlone       antonio carlo carlone  
...                            ...                         ...  
461        corrado e giulio venini     corrado e giulio venini  
460                         veglia                      veglia  
459                         varese                      varese  
467                       viminale                    viminale  
1380                        nirone                      nirone  

[1381 rows x 8 columns]

In [39]:
# Function to compute Jaccard similarity
def compute_jaccard(s1, s2):
    vectorizer = CountVectorizer().fit_transform([s1, s2])
    vectors = vectorizer.toarray()
    return 1 - jaccard(vectors[0], vectors[1])

# Function to compute Jaccard similarity for all rows in viario and update non_matching_df
def compute_jaccard_for_all_rows():
    for index, row in tqdm(check_matching_df[check_matching_df.similarity < 77].iterrows(), total=len(check_matching_df[check_matching_df.similarity < 77])):
        normalized_ubic_via = row['normalized_ubic_via']
        
        # Compute Jaccard similarity for 'descrittivo' and 'denominazione' for all rows in viario
        viario['jaccard_descrittivo'] = viario['normalized_descrittivo'].apply(
            lambda x: compute_jaccard(normalized_ubic_via, x))
        viario['jaccard_denominazione'] = viario['normalized_denominazione'].apply(
            lambda x: compute_jaccard(normalized_ubic_via, x))
        
        # Take the maximum Jaccard similarity (between 'descrittivo' and 'denominazione')
        viario['jaccard_similarity'] = viario[['jaccard_descrittivo', 'jaccard_denominazione']].max(axis=1)
        
        # Find the best match based on the highest Jaccard similarity
        best_match = viario.loc[viario['jaccard_similarity'].idxmax()]
        
        # Update the non_matching_df with the best match and Jaccard similarity
        check_matching_df.at[index, 'best_normalized_descrittivo'] = best_match['normalized_descrittivo']
        check_matching_df.at[index, 'best_normalized_denominazione'] = best_match['normalized_denominazione']
        check_matching_df.at[index, 'best_descrittivo'] = best_match['DESCRITTIVO']
        check_matching_df.at[index, 'best_denominazione'] = best_match['DENOMINAZIONE']
        check_matching_df.at[index, 'measure'] = "JACCARD"
        check_matching_df.at[index, 'similarity'] = best_match['jaccard_similarity'] * 100

# Run Jaccard computation for all rows in viario and update non_matching_df
compute_jaccard_for_all_rows()

# Show the updated non_matching_df with Jaccard similarity values
print(f"Updated non_matching_df with Jaccard similarity:")
check_matching_df.sort_values(by="similarity")

100%|██████████| 7/7 [00:29<00:00,  4.20s/it]

Updated non_matching_df with Jaccard similarity:


measure  similarity                 ubic_via       best_denominazione  \
183   JACCARD         0.0                                       DUOMO (DEL)   
292   JACCARD        50.0    COLONNA MARCO ANTONIO  BRAGADINO MARCO ANTONIO   
1172  JACCARD        50.0      GRAZIANO IMPERATORE                 GRAZIANO   
1265  JACCARD        50.0          DANTE ALIGHIERI                    DANTE   
1103  JACCARD        50.0         TIZIANO VECELLIO                 TIZIANO    
...       ...         ...                      ...                      ...   
461      EDIT       100.0  VENINI GIULIO E CORRADO  VENINI GIULIO E CORRADO   
460      EDIT       100.0                   VEGLIA                   VEGLIA   
459      EDIT       100.0                   VARESE                   VARESE   
467      EDIT       100.0                 VIMINALE                 VIMINALE   
1380     EDIT       100.0                   NIRONE                   NIRONE   

             best_descrittivo      normalized_ubic_via  \
183                 DEL DUOMO                            
292   MARCO ANTONIO BRAGADINO    antonio colonna marco   
1172                 GRAZIANO      graziano imperatore   
1265                    DANTE          alighieri dante   
1103                  TIZIANO         tiziano vecellio   
...                       ...                      ...   
461   GIULIO E CORRADO VENINI  corrado e giulio venini   
460                    VEGLIA                   veglia   
459                    VARESE                   varese   
467                  VIMINALE                 viminale   
1380                   NIRONE                   nirone   

     best_normalized_denominazione best_normalized_descrittivo  
183                      del duomo                   del duomo  
292        antonio bragadino marco     antonio bragadino marco  
1172                      graziano                    graziano  
1265                         dante                       dante  
1103                       tiziano                     tiziano  
...                            ...                         ...  
461        corrado e giulio venini     corrado e giulio venini  
460                         veglia                      veglia  
459                         varese                      varese  
467                       viminale                    viminale  
1380                        nirone                      nirone  

[1381 rows x 8 columns]

With the exception of two rows, which could not be recognized due to missing values in the `ubic_via` column, most of the rows now exhibit a similarity of 50% or greater, indicating that their addresses match semantically. As a result, the `ubic_via` column can now be standardized to align with the format of the `viario.DESCRITTIVO` column.

In [40]:
# Step 1: Merge new_SERVICES with check_matching_df on 'ubic_via'
merged_df = new_SERVICES.merge(check_matching_df[['ubic_via', 'best_descrittivo', 'similarity']], 
                               on='ubic_via', how='left')

# Step 2: Filter rows where similarity >= 50
filtered_df = merged_df[merged_df['similarity'] >= 50]

# Step 3: Update the 'ubic_via' column in new_SERVICES with 'DENOMINAZIONE' from check_matching_df where similarity >= 50
for index, row in filtered_df.iterrows():
    new_SERVICES.loc[new_SERVICES['ubic_via'] == row['ubic_via'], 'ubic_via'] = row['best_descrittivo']

new_SERVICES.head()

t_es   t_via                via civ  cod_via zd  sup_alt  \
0               NaN   Largo      DEI GELSOMINI  10     5394  6      NaN   
1   Centro Massaggi  Piazza              FIDIA   3     1144  9      2.0   
2  Centro Benessere     Via              ADIGE  10     4216  5      2.0   
3        Truccatore     Via  FLAVIO BARACCHINI   9      356  1      NaN   
4               NaN     Via            BERGAMO  12     3189  4      NaN   

   sup_lav ubic_t_via           ubic_via ubic_civ ubic_zd note  
0     55.0      Largo      DEI GELSOMINI       10       6  NaN  
1     28.0     Piazza              FIDIA        3       9  NaN  
2     27.0        Via              ADIGE       10       5  NaN  
3      NaN        Via  FLAVIO BARACCHINI        9       1  NaN  
4     50.0        Via            BERGAMO       12       4  NaN

# Error Detection and Correction
## UBIC-VIA: Solve Mismatches

The goal is to resolve mismatches between the `ubic` columns (`ubic_t_via`, `ubic_via`, `ubic_civ`, `ubic_zd`) and the corresponding main columns (`t_via`, `via`, `civ`, `zd`). Mismatches are identified by comparing values in these columns. 

Once detected, corrections are made by updating the values to ensure consistency, improving the dataset's quality and accuracy for analysis.

In [41]:
# Initialize dictionary to store mismatches
mismatch_rows = {
    "note empty": [],
    "t_via not match": [],
    "civ not match": [],
    "via not match": [],
    "zd not match": [],
}

# Function to check conditions and log mismatched rows
def check_and_log_mismatches(new_SERVICES):
    # Create masks for each condition (True if mismatch, False if match)
    note_not_empty_condition = ~new_SERVICES['note'].isna()
    t_via_not_match_condition = new_SERVICES["t_via"] != new_SERVICES["ubic_t_via"]
    civ_not_match_condition = new_SERVICES["civ"] != new_SERVICES["ubic_civ"]
    via_not_match_condition = new_SERVICES["via"] != new_SERVICES["ubic_via"]
    zd_not_match_condition = new_SERVICES["zd"] != new_SERVICES["ubic_zd"]
    
    # Apply the conditions to filter rows and store them in the corresponding lists
    mismatch_rows["note empty"] = new_SERVICES[note_not_empty_condition].to_dict(orient='records')
    mismatch_rows["t_via not match"] = new_SERVICES[t_via_not_match_condition].to_dict(orient='records')
    mismatch_rows["civ not match"] = new_SERVICES[civ_not_match_condition].to_dict(orient='records')
    mismatch_rows["via not match"] = new_SERVICES[via_not_match_condition].to_dict(orient='records')
    mismatch_rows["zd not match"] = new_SERVICES[zd_not_match_condition].to_dict(orient='records')
    
    return new_SERVICES

# Apply the function to check and log mismatches
new_SERVICES = check_and_log_mismatches(new_SERVICES)

# Print summary of mismatches with aligned output
print("Summary of mismatches:\n")
# Print a header row with column names
print(f"{'Mismatch Type':<20} {'Number of Mismatches':<20}")
print("="*40)  # Separator for clarity

# Iterate over mismatch_rows and print each mismatch type and its count
for key, value in mismatch_rows.items():
    print(f"{key:<20} \t\t{len(value):<20}")

Summary of mismatches:

Mismatch Type        Number of Mismatches
note empty           		399                 
t_via not match      		20                  
civ not match        		151                 
via not match        		76                  
zd not match         		17                  


Check if the full `ubic` and `via_full` addresses are present in the dataset of **civici**

In [42]:
new_SERVICES['zd'] = new_SERVICES['zd'].astype(int)

def check_full_via_exists(row, civici_df):
    # Check if the necessary columns contain NaN values
    if pd.isna(row['ubic_t_via']) or pd.isna(row['ubic_via']) or pd.isna(row['t_via']) or pd.isna(row['via']):
        return False  # Return False if any of the critical columns have NaN values
    
    # Combine the address components if valid
    full_via_ubic = row['ubic_t_via'] + " " + row['ubic_via']
    full_via_via = row['t_via'] + " " + row['via']
    
    # Check if full_via_ubic exists in civici.DESCRITTIVO and match NUMEROCOMPLETO and MUNICIPIO
    match_ubic = civici_df[(civici_df['DESCRITTIVO'] == full_via_ubic) &
                            (civici_df['NUMEROCOMPLETO'] == row['ubic_civ']) &
                            (civici_df['MUNICIPIO'] == row['ubic_zd'])]
    
    # Check if full_via_via exists in civici.DESCRITTIVO and match NUMEROCOMPLETO and MUNICIPIO
    match_via = civici_df[(civici_df['DESCRITTIVO'] == full_via_via) &
                           (civici_df['NUMEROCOMPLETO'] == row['civ']) &
                           (civici_df['MUNICIPIO'] == row['zd'])]
    
    # Return True if either match is found, else False
    return True if len(match_ubic) > 0 or len(match_via) > 0 else False

# Use tqdm to track the progress of the apply function
tqdm.pandas()

# Apply the check function to each row in your dataset and create two flags
new_SERVICES[['ubic_flag', 'full_via_flag']] = new_SERVICES.progress_apply(
    lambda row: pd.Series({
        'ubic_flag': check_full_via_exists(row, civici),
        'full_via_flag': check_full_via_exists(row, civici)
    }), axis=1
)

# Show the result
new_SERVICES[['ubic_flag', 'full_via_flag']].head()


# Show the result
new_SERVICES.head()

100%|██████████| 3908/3908 [02:37<00:00, 24.88it/s]


t_es   t_via                via civ  cod_via  zd  sup_alt  \
0               NaN   Largo      DEI GELSOMINI  10     5394   6      NaN   
1   Centro Massaggi  Piazza              FIDIA   3     1144   9      2.0   
2  Centro Benessere     Via              ADIGE  10     4216   5      2.0   
3        Truccatore     Via  FLAVIO BARACCHINI   9      356   1      NaN   
4               NaN     Via            BERGAMO  12     3189   4      NaN   

   sup_lav ubic_t_via           ubic_via ubic_civ ubic_zd note  ubic_flag  \
0     55.0      Largo      DEI GELSOMINI       10       6  NaN       True   
1     28.0     Piazza              FIDIA        3       9  NaN       True   
2     27.0        Via              ADIGE       10       5  NaN       True   
3      NaN        Via  FLAVIO BARACCHINI        9       1  NaN       True   
4     50.0        Via            BERGAMO       12       4  NaN       True   

   full_via_flag  
0           True  
1           True  
2           True  
3           True  
4           True

In [43]:
new_SERVICES[(new_SERVICES.ubic_flag == False) | (new_SERVICES.full_via_flag == False) ]

t_es  \
143                                                                                                             Acconciatore   
163                                                                                                             Acconciatore   
209                                                                                                             Acconciatore   
242                                                                                                             Acconciatore   
285                                                                                                             Acconciatore   
...                                                                                                                      ...   
3717                                                               Tipo B - Centro di Abbronzatura;Tipo A - Estetica Manuale   
3740                                                               Tipo B - Centro di Abbronzatura;Tipo A - Estetica Manuale   
3800                                                               Tipo B - Centro di Abbronzatura;Tipo A - Estetica Manuale   
3838                                                                                         Tipo B - Centro di Abbronzatura   
3906  Tipo D - Estetica Apparati Elettromeccanici;Tipo C - Trattamenti Estetici Dimagrimento;Tipo B - Centro di Abbronzatura   

      t_via                  via  civ  cod_via  zd  sup_alt  sup_lav  \
143     Via                PAVIA  NaN     5262   5      NaN      NaN   
163     Via  SALA DEI LONGOBARDI    2      217   1      NaN      NaN   
209   Corso                 LODI   72     4068   4      NaN     50.0   
242   Corso             SEMPIONE   10     7137   1      NaN      NaN   
285     Via            APPENNINI  NaN     7514   8      NaN      NaN   
...     ...                  ...  ...      ...  ..      ...      ...   
3717    Via     LUIGI PIRANDELLO  NaN     6242   7      NaN    100.0   
3740    Via  SALA DEI LONGOBARDI    2      217   1      NaN      NaN   
3800  Viale              CERTOSA  NaN     7174   8      NaN      NaN   
3838    Via                 BARI  NaN     5218   6      NaN     29.0   
3906    Via               NIRONE  NaN      640   1      NaN      NaN   

     ubic_t_via             ubic_via ubic_civ ubic_zd                    note  \
143         Via                PAVIA        6       5                      /2   
163         Via  SALA DEI LONGOBARDI        2       1                     NaN   
209         NaN                             8       4            codvia 4386    
242         NaN                             0       1  mm1 duomo codvia 9112    
285         Via            APPENNINI      77H       8                     NaN   
...         ...                  ...      ...     ...                     ...   
3717        Via     LUIGI PIRANDELLO       4A       7     piano seminterrato    
3740        Via  SALA DEI LONGOBARDI        2       1                     NaN   
3800      Viale              CERTOSA      21A       8       interno palestra    
3838        Via                 BARI      11D       6                     NaN   
3906        Via               NIRONE       2A       1                     NaN   

      ubic_flag  full_via_flag  
143       False          False  
163       False          False  
209       False          False  
242       False          False  
285       False          False  
...         ...            ...  
3717      False          False  
3740      False          False  
3800      False          False  
3838      False          False  
3906      False          False  

[85 rows x 15 columns]

In [44]:
new_SERVICES[(new_SERVICES.ubic_flag == False) | (new_SERVICES.full_via_flag == False)].to_csv("unmatched.csv")

## We start the cleaning (to fix) -> actually we start fixing the values (maybe to put before?)
<!-- **Assumption**: `The "codice via" value is correct`. <br> -->
Now per each column we are going to study the mismatches and clean the data.

**ZD mismatches**: we use the dataset VIARIO provided by the municipality of Milan to assess the true values. <br>
**CIVIC mismatches** we use the dataset CIVICI provided by the municipality of Milan to assess the true values. In case there is no equal values we assume the "civ" value true. <br>
**VIA mismatches** we use the information in the column "Note" if provided. Otherwise we drop the row.

#### ZD mismatches
We are going to use as **ground truth** our VIARIO_SECOND dataset using the `cod_via` as key.

In [45]:
# # Function to normalize the names by removing extra spaces, converting to lowercase, and sorting words
# def normalize_string(s):

#     #to remove specific pattern contained in the data
#     s = re.sub(r"sant'\s*", "sant'", s.lower())  # Fix spaces around 'sant'
#     s = re.sub(r"dell'\s*", "dell'", s.lower())  # Fix spaces around 'dell
#     s = re.sub(r"[()]", "", s) # Remove the parenthesis
    
#     # Split the string into words
#     words = s.strip().split()
    
#     # Sort the words alphabetically
#     words.sort()
    
#     # Join the words back together and convert to lowercase
#     return " ".join(words)

# # Function to compute the edit distance between two strings (after normalization)
# def calculate_similarity(str1, str2):
#     return fuzz.ratio(str1, str2)

# # Merge new_SERVICES with viario on 'cod_via' and 'CODICE_VIA' to bring in 'DESCRITTIVO'
# merged_df = new_SERVICES.merge(viario[['CODICE_VIA', 'DESCRITTIVO', 'DENOMINAZIONE']], left_on='cod_via', right_on='CODICE_VIA', how='left')

# # Normalize 'via' and 'DESCRITTIVO' columns
# merged_df['normalized_via'] = merged_df['via'].apply(normalize_string)
# merged_df['normalized_descrittivo'] = merged_df['DESCRITTIVO'].apply(normalize_string)
# merged_df['normalized_denominazione'] = merged_df['DENOMINAZIONE'].apply(normalize_string)

# # Calculate the similarity between normalized values using edit distance
# merged_df['similarity'] = merged_df.apply(lambda row: max(calculate_similarity(row['normalized_via'], row['normalized_descrittivo']), calculate_similarity(row['normalized_via'], row['normalized_denominazione'])), axis=1)

# # Print rows where the similarity is below a certain threshold (e.g., 80)
# threshold = 75
# non_matching_rows = merged_df[merged_df['similarity'] < threshold]

# # Show the rows with low similarity
# print(f"Rows where normalized 'via' and 'DESCRITTIVO' are not similar enough (below {threshold}% similarity):")
# non_matching_rows[['cod_via', 'via', 'DENOMINAZIONE', 'DESCRITTIVO', 'normalized_via', 'normalized_descrittivo', 'normalized_denominazione', 'similarity']].sort_values(by="similarity", ascending=True)

In [46]:
# # Step 1: Merge the DataFrames on cod_via and CODICE_VIA
# merged_df = new_SERVICES.merge(viario_centroidi[['CODICE_VIA', 'MUNICIPIO']], left_on='cod_via', right_on='CODICE_VIA', how='left')

# merged_df['zd'] = merged_df['zd'].astype(int)

# # Step 2: Compare new_SERVICES.zd with viario.MUNICIPIO
# merged_df['zd_match'] = merged_df['zd'] == merged_df['MUNICIPIO']

# # Step 3: Extract rows where zd and MUNICIPIO do not match
# mismatch_rows = merged_df[~merged_df['zd_match']]

# # Step 4: Display the mismatched rows for review
# print("Mismatched rows between 'zd' and 'MUNICIPIO':")
# print(mismatch_rows[['cod_via', 'via', 'zd', 'MUNICIPIO']])

In [47]:
# # Get unique values
# unique_values = civici['NUMEROCOMPLETO'].unique()

# # Sort the unique values
# sorted_values = sorted(unique_values)

# # Display the sorted values
# print(sorted_values)

# To start again from here

In [48]:
# zd_condition = (new_SERVICES["zd"]!=new_SERVICES["ubic_zd"])

In [49]:
# new_SERVICES[zd_condition].head()

In [50]:
# new_SERVICES.head()

In [51]:
# civici_milan.DESCRITTIVO.unique()

In [52]:
# def check_zd_in_viario(row, viario_df):
#     """ Function to check and update 'zd' and 'ubic_zd' based on 'viario_df' dataset """
    
#     ubic_zd = row.get('ubic_zd', None) 
#     zd = row.get('zd', None) 
#     cod_via = row.get('cod_via', None)  # Extract 'cod_via' from the row
    
#     # Find the corresponding MUNICIPIO in viario_df using cod_via
#     viario_row = viario_df[viario_df['CODICE_VIA'] == cod_via]
    
#     if not viario_row.empty:
#         zd_viario = viario_row['MUNICIPIO'].values[0]  # Get the MUNICIPIO value
#     else:
#         zd_viario = None  # If no matching cod_via found, set zd_viario to None
    
#     # Check if MUNICIPIO matches ubic_zd or zd and update accordingly
#     if str(zd_viario) == str(ubic_zd):
#         row['zd'] = ubic_zd
#     elif str(zd_viario) == str(zd):
#         row['ubic_zd'] = zd
#     else:
#         # Default to municipio for both if no match
#         row['ubic_zd'] = zd_viario
#         row['zd'] = zd_viario
    
#     return row

# def update_rows_with_viario(new_services, viario_second, zd_condition):
    
#     # Apply the function to each row of new_SERVICES[zd_condition]
#     updated_rows = new_services[zd_condition].apply(check_zd_in_viario, axis=1, viario_df=viario_second)
    
#     # Replace the updated rows in new_SERVICES
#     new_services.loc[zd_condition] = updated_rows
    
#     return new_services

# new_SERVICES = update_rows_with_viario(new_SERVICES, viario_second, zd_condition)

### Civico
We are going to use the "CiviciCdMi" dataset provided by the municipality of Milan on their website: https://geoportale.comune.milano.it/sit/toponomastica/ as ground truth.

In [53]:
# def normalize_column(column):
#     return (
#         column.fillna("")  # Replace NaN with empty strings
#         .astype(str)       # Ensure the column is of string type
#         .str.strip()       # Remove leading and trailing whitespaces
#         .str.lower()       # Normalize to lowercase for comparison
#     )

# new_SERVICES["civ"] = normalize_column(new_SERVICES["civ"])
# new_SERVICES["ubic_civ"] = normalize_column(new_SERVICES["ubic_civ"])

# # Recompute the condition
# civ_condition = (new_SERVICES["civ"] != new_SERVICES["ubic_civ"])

In [54]:
# civ_condition_to_drop = pd.isna(new_SERVICES["civ"]) & pd.isna(new_SERVICES["ubic_civ"])

In [55]:
# new_SERVICES = new_SERVICES[~civ_condition_to_drop]

In [56]:
# new_SERVICES[civ_condition].shape

In [57]:
# def check_civ_in_civici(row, civici_df):
#     """ Function to check and update 'civ' and 'ubic_civ' based on 'viario_df' dataset """
    
#     # Get values from the row
#     ubic_civ = row.get('ubic_civ', None) 
#     civ = row.get('civ', None) 
#     cod_via = row.get('cod_via', None)  
    
#     # Handle NaN values explicitly, treat them as empty strings
#     if pd.isna(ubic_civ):
#         ubic_civ = ''
#     if pd.isna(civ):
#         civ = ''
    
#     # Get the rows from civici_df that match the 'cod_via'
#     civici_rows = civici_df[civici_df['CODICE_VIA'] == cod_via]

#     row['valid'] = True

#     if not civici_rows.empty:
#         arr_civici = civici_rows['NUMEROCOMPLETO'].tolist()  # Collect all 'NUMEROCOMPLETO' values
                
#         if civ in arr_civici and ubic_civ not in arr_civici:
#             # Only 'civ' is valid, update 'ubic_civ' with 'civ'
#             row['ubic_civ'] = civ
#         elif ubic_civ in arr_civici and civ not in arr_civici:
#             # Only 'ubic_civ' is valid, update 'civ' with 'ubic_civ'
#             row['civ'] = ubic_civ
#         else:
#             # Neither 'ubic_civ' nor 'civ' are valid
#             row[:] = pd.NA
#     else:
#         # No matching rows in civici_df, mark as invalid
#         row[:] = pd.NA
        
#     return row

In [58]:
# def update_rows_with_civici(new_services, civici_milan, civ_condition):
    
#     # Apply the function to each row of new_SERVICES[zd_condition]
#     updated_rows = new_services[civ_condition].apply(check_civ_in_civici, axis=1, civici_df=civici_milan)
    
#     # Replace the updated rows in new_SERVICES
#     new_services.loc[civ_condition] = updated_rows

#     new_services = new_services.dropna(how='all')
    
#     return new_services

# new_SERVICES = update_rows_with_civici(new_SERVICES, civici_milan, civ_condition)

### Via
We are going to study the data in the "Note" column. The idea is to retrieve the via names stored in this column. In case this value is equal to either one between "via" and "ubic_via" that one will be considered as corrected. 

In [59]:
# via_condition = (new_SERVICES["via"]!=new_SERVICES["ubic_via"])
# note_condition = (new_SERVICES["note"]!="")

In [60]:
# new_SERVICES[via_condition].shape

In [61]:
# def preprocess_note(note):
#     # Convert to lower case for case-insensitive matching
#     note = str(note).lower()
    
#     # Check if the note contains 'via' or 'v.le'
#     if "via" in note:
#         # Keep only the part after 'via'
#         note = note.split("via", 1)[1].strip()  # Keep everything after 'via'
#     elif "v.le" in note:
#         # Keep only the part after 'v.le'
#         note = note.split("v.le", 1)[1].strip()  # Keep everything after 'v.le'
#     else:
#         # Return an empty string if it doesn't contain 'via' or 'v.le'
#         note = ""
    
#     return note

# # Apply the preprocess function to the 'note' column
# new_SERVICES.loc[via_condition, 'note'] = new_SERVICES.loc[via_condition, 'note'].apply(preprocess_note)

In [62]:
# new_SERVICES.loc[via_condition & note_condition].head()

There are only 3 rows where the via and ubic_via are different and the processed note value is not null. Despite that the note doesn't add any type of information. <br>
We proceed in removing all the rows with a mismatch between "via" and "ubic_via"

In [63]:
# new_SERVICES = new_SERVICES.loc[~via_condition]

In [64]:
# # Initialize dictionary to store mismatches
# mismatch_rows = {
#     "note not empty": [],
#     "civ not match": [],
#     "via not match": [],
#     "zd not match": [],
# }

# # Apply the function to check and log mismatches
# new_SERVICES = check_and_log_mismatches(new_SERVICES)

# # Print summary of mismatches
# print("\nSummary of mismatches:")
# for key, value in mismatch_rows.items():
#     print(f"{key}: {len(value)} mismatches")

We can also drop the column "Note" since it doesn't add any meaningful information

In [65]:
# new_SERVICES = new_SERVICES.drop(columns=["note"])

In [66]:
#Qui dentro si potranno visualizzare meglio le colonne.
# new_SERVICES.to_csv("new_final.csv")

# Removal (to decide what to do)
We drop rows labeled as "Truccatore" in "t_es", since the class is under-represented.

In [67]:
# new_SERVICES = new_SERVICES[new_SERVICES["t_es"] != "Truccatore"]